core/Enclave/enclave_hsm.cpp

```diff
@@ -43,767 +42,641 @@
-        gcm_data->aad_size = real_aad_size;
-    }
-
-    return ret;
-}
-
-static sgx_status_t sgx_gcm_decrypt(const sgx_aes_gcm_128bit_key_t *key,
-                                    uint32_t *plaintext_size, uint8_t *plaintext,
-                                    const sgx_aes_gcm_data_ex_t *gcm_data)
-{
-    if (NULL == gcm_data || NULL == plaintext || NULL == *plaintext_size
-                 || *plaintext_size < sgx_get_gcm_ciphertext_size(gcm_data))
+    if (cmk->metadata.keyspec >= INVALID_VALUE)
+    {
         return SGX_ERROR_INVALID_PARAMETER;
-
-    sgx_status_t ret = sgx_rijndael128GCM_decrypt(key,
-                                                  gcm_data->payload, gcm_data->ciphertext_size,
-                                                  plaintext,
-                                                  gcm_data->iv, sizeof(gcm_data->iv),
-                                                  &(gcm_data->payload[gcm_data->ciphertext_size]), gcm_data->aad_size,
-                                                  (const sgx_aes_gcm_128bit_tag_t*)gcm_data->mac);
-    if (SGX_SUCCESS != ret)
-        printf("gcm decrypting failed.\n");
-    else
-        *plaintext_size = sgx_get_gcm_ciphertext_size(gcm_data);
-
-    return ret;
-}
-
-
-sgx_status_t enclave_create_aes_key(uint8_t *cmk_blob, uint32_t cmk_blob_size, uint32_t *req_blob_size)
-{
-    sgx_status_t ret = SGX_ERROR_UNEXPECTED;
-    uint32_t real_blob_len = sgx_calc_gcm_data_size(0, SGX_AES_KEY_SIZE);
-
-    if (real_blob_len == UINT32_MAX)
-        return SGX_ERROR_UNEXPECTED;
-
-    if (req_blob_size != NULL) {
-        *req_blob_size = real_blob_len;
-        return SGX_SUCCESS;
     }
-
-    if (cmk_blob == NULL || cmk_blob_size != real_blob_len)
-        return SGX_ERROR_INVALID_PARAMETER;
-
-    uint8_t* tmp = (uint8_t *)malloc(SGX_AES_KEY_SIZE);
-    if (tmp == NULL)
-        return SGX_ERROR_OUT_OF_MEMORY;
-
-    ret = sgx_read_rand(tmp, SGX_AES_KEY_SIZE);
-    if (ret != SGX_SUCCESS) {
-        free(tmp);
-        return ret;
+    switch (cmk->metadata.keyspec)
+    {
+    case EH_AES_GCM_128:
+    case EH_AES_GCM_192:
+    case EH_AES_GCM_256:
+        if (cmk->keybloblen == 0)
+        {
+            ret = ehsm_create_aes_key(NULL,
+                                      0,
+                                      &(cmk->keybloblen),
+                                      (ehsm_keyspec_t)(cmk->metadata.keyspec));
+        }
+        else
+        {
+            ret = ehsm_create_aes_key(cmk->keyblob,
+                                      cmk->keybloblen,
+                                      NULL,
+                                      (ehsm_keyspec_t)(cmk->metadata.keyspec));
+        }
+        break;
+    case EH_RSA_2048:
+    case EH_RSA_3072:
+    case EH_RSA_4096:
+    case EH_EC_P224:
+    case EH_EC_P256:
+    case EH_EC_P384:
+    case EH_EC_P512:
+    case EH_SM2:
+        ret = ehsm_create_asymmetric_key(cmk);
```

Comment:
why still not split this function?

Reply:
this has been changed into ehsm_create_rsa_key and ehsm_create_ec_key